### PR TITLE
Allow specification of resource requirements in tools.

### PR DESCRIPTION
**Pull request by jmchilton**
Backend doesn't yet support them (except for providing a list of `ResourceRequirement` on the tool instance, so dynamic job rules could use them), but allow tool authors to specify what the runtime should support. Based on CWL resource requirements (http://www.commonwl.org/v1.2/CommandLineTool.html#ResourceRequirement).

xref https://github.com/galaxyproject/galaxy/issues/5369

### CODE REVIEW

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 1]
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
       <xs:element name="requirement" type="Requirement" minOccurs="0" maxOccurs="unbounded"/>
       <xs:element name="container" type="Container" minOccurs="0" maxOccurs="unbounded"/>
     </xs:sequence>
+    <xs:attribute name="cores_min" type="xs:long" use="required">
```
Comment (bernt-matthias) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]: `use="required"` is making it mandatory to specify a value if the attribute is set, or?

**Comment by mr-c** [comment]
Also include ToolTimeLimit https://www.commonwl.org/v1.1/CommandLineTool.html#ToolTimeLimit ?

File: lib/galaxy/tool_util/xsd/galaxy.xsd (hunk 1 above)
Comment (mvdbeek) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
    <xs:attribute name="cores_min" type="xs:long" use="optional">
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 2]
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
...
+      <xs:documentation xml:lang="en"><![CDATA[
+Maximum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="cores_max" type="xs:long" use="required">
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
    <xs:attribute name="cores_max" type="xs:long" use="optional">
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 3]
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
...
+    </xs:attribute>
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    <xs:attribute name="ram_min" type="xs:long" use="required">
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
    <xs:attribute name="ram_min" type="xs:long" use="optional">
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 4]
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
...
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved RAM in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="ram_max" type="xs:long" use="required">
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
    <xs:attribute name="ram_max" type="xs:long" use="optional">
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 5]
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
...
+      <xs:documentation xml:lang="en"><![CDATA[
+Maximum reserved RAM in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="tmpdir_min" type="xs:long" use="required">
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
    <xs:attribute name="tmpdir_min" type="xs:long" use="optional">
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 6]
```diff
@@ -322,6 +322,36 @@ serve as complete descriptions of the runtime of a tool.
...
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved filesystem based storage for the designated temporary directory, in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="tmpdir_max" type="xs:long" use="required">
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
    <xs:attribute name="tmpdir_max" type="xs:long" use="optional">
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 7]
```diff
@@ -614,6 +614,36 @@ serve as complete descriptions of the runtime of a tool.
       <xs:element name="requirement" type="Requirement" minOccurs="0" maxOccurs="unbounded"/>
       <xs:element name="container" type="Container" minOccurs="0" maxOccurs="unbounded"/>
     </xs:sequence>
+    <xs:attribute name="cores_min" type="xs:long" use="optional">
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]: Looks like xs:long isn't a valid type, but the [CWL 1.2 spec](https://www.commonwl.org/v1.2/CommandLineTool.html#ResourceRequirement) allows this to be floats, so maybe we should go with
```suggestion
    <xs:attribute name="cores_min" type="float" use="optional">
```
?

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 8]
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
...
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="cores_max" type="xs:float" use="optional">
+    </xs:attribute>
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    <xs:attribute name="ram_min" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved RAM in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
Minimum reserved RAM in mebibytes (2**20 bytes), if runtime allows it (not yet implemented in Galaxy).
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 9]
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
...
     </xs:sequence>
+    <xs:attribute name="cores_min" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Maximum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="cores_max" type="xs:float" use="optional">
+    </xs:attribute>
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
Maximum reserved number of CPU cores, if runtime allows it (not yet implemented in Galaxy).
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 10]
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
       <xs:element name="requirement" type="Requirement" minOccurs="0" maxOccurs="unbounded"/>
       <xs:element name="container" type="Container" minOccurs="0" maxOccurs="unbounded"/>
     </xs:sequence>
+    <xs:attribute name="cores_min" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Maximum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
Minimum reserved number of CPU cores, if runtime allows it (not yet implemented in Galaxy).
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 11]
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
...
+Minimum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    <xs:attribute name="ram_min" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved RAM in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="ram_max" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Maximum reserved RAM in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
Maximum reserved RAM in mebibytes (2**20 bytes), if runtime allows it (not yet implemented in Galaxy).
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 12]
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
...
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="tmpdir_min" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved filesystem based storage for the designated temporary directory, in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
Minimum reserved filesystem-based storage for the designated temporary directory in mebibytes (2**20 bytes), if runtime allows it (not yet implemented in Galaxy).
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 13]
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
...
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="tmpdir_max" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Maximum reserved filesystem based storage for the designated temporary directory, in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
Maximum reserved filesystem-based storage for the designated temporary directory in mebibytes (2**20 bytes), if runtime allows it (not yet implemented in Galaxy).
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 14]
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
       <xs:element name="requirement" type="Requirement" minOccurs="0" maxOccurs="unbounded"/>
       <xs:element name="container" type="Container" minOccurs="0" maxOccurs="unbounded"/>
     </xs:sequence>
+    <xs:attribute name="cores_min" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Maximum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="cores_max" type="xs:float" use="optional">
+    </xs:attribute>
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 15]
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
...
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved number of CPU cores if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
]]></xs:documentation>
    </xs:attribute>
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 16]
```diff
@@ -614,6 +614,37 @@ serve as complete descriptions of the runtime of a tool.
...
+      <xs:documentation xml:lang="en"><![CDATA[
+Minimum reserved filesystem based storage for the designated temporary directory, in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+    <xs:attribute name="tmpdir_max" type="xs:float" use="optional">
+      <xs:documentation xml:lang="en"><![CDATA[
+Maximum reserved filesystem based storage for the designated temporary directory, in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).
+]]></xs:documentation>
+    </xs:attribute>
+  </xs:attribute>
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd (hunk 10 above)
Comment (mvdbeek) [reply lib/galaxy/tool_util/xsd/galaxy.xsd]: I'm redoing this, I don't think we'd want to write javascript expressions within attributes.
I'm thinking of this:
```xml
    <requirements>
        <resource type="cores_min">1.1</resource>
        <resource type="ram_max"><![CDATA[runtime.cores * 2]]></resource>
    </requirements>
```

**Comment by nuwang** [comment]
Glad to see all the progress on this PR. @Slugger70, @cat-bro and I discussed this PR's new resource requirements in relation to TPV and we had some questions:
a) How do you envision the min and max resource requirements working in practice? In the case of SLURM for example, it needs to translate into a native spec with a specific core and mem count.
b) Is it possible to capture a single minimum requirement value independent of input size?
c) Would the maximum value be too dependent on actual resource availability, assuming the tool always scales with greater resources. In other words, would these values be too site specific, and not generally specifiable?
d) How should admins use these values with existing dynamic rules? (in TPV or DTDs for example). Especially in relation to conditional scaling rules.

Once we have a better handle on this, the plan was to modify TPV accordingly with support for the same constructs.

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 17]
```diff
@@ -6724,6 +6746,43 @@ and ``bibtex`` are the only supported options.</xs:documentation>
...
+        </xs:annotation>
+      </xs:enumeration>
+      <xs:enumeration value="tmpdir_min">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">Minimum reserved filesystem-based storage for the designated temporary directory in mebibytes (2**20 bytes), if runtime allows it (unimplemented in Galaxy currently).</xs:documentation>
+        </xs:annotation>
+      </xs:enumeration>
+      <xs:enumeration value="tmpdir_max">
+        <xs:annotation>
+          <xs:documentation xml:lang="en">Maximum reserved filesystem based storage for the designated temporary directory, in mebibytes (2**20) if runtime allows it (unimplemented in Galaxy currently).</xs:documentation>
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
          <xs:documentation xml:lang="en">Maximum reserved filesystem based storage for the designated temporary directory in mebibytes (2**20 bytes), if runtime allows it (unimplemented in Galaxy currently).</xs:documentation>
```

**Review by jdavcs — APPROVED** [review]
LGTM, except edits by @nsoranzo cannot be automatically applied - they should be included before we merge this.

**Comment by mvdbeek** [comment]
Did I miss something ? I did apply them locally

**Comment by jdavcs** [comment]
The proposed edit was to replace "unimplemented" with "not yet implemented"
Here's the proposed edit: https://github.com/galaxyproject/galaxy/pull/5378#discussion_r897119178
Here's the current file: https://github.com/galaxyproject/galaxy/blob/2c65a12c0b3df05dcae9458d41f8d19df754aef9/lib/galaxy/tool_util/xsd/galaxy.xsd#L6771
At least that's what's on my screen?

**Comment by jdavcs** [comment]
I see your previous commit. Did I misread the proposed edit? Was it the other way round? If so, sorry for the confusion!

**Comment by mvdbeek** [comment]
I see, I changed it twice then

**Comment by mr-c** [comment]
> > c) Would the maximum value be too dependent on actual resource availability, assuming the tool always scales with greater resources. In other words, would these values be too site specific, and not generally specifiable?
> 
> I think we're free to overallocate resources (maybe not for CWL, ping @mr-c, but definitely for Galaxy)

Yep, over-resourcing is fine; hopefully the actual resources & and highest usage are recorded in the provenance

**Comment by nsoranzo** [comment]
> I see, I changed it twice then

Yes, 2c65a12c0b3df05dcae9458d41f8d19df754aef9 is undoing what I suggested. No big deal any way.

File: test/functional/tools/resource_requirements.xml [hunk 18]
```diff
@@ -0,0 +1,19 @@
+<tool id="resource_requirements" name="resource_requirements" version="1.0.0" profile="22.05">
+    <requirements>
+        <resource type="cores_min">1.1</resource>
```
Comment (nsoranzo) [inline test/functional/tools/resource_requirements.xml:3]: Here you're using `resource` while in other places the element is `resource_requirement`. I think `resource` may be better (shorter and already inside `requirements`).

File: lib/galaxy/tool_util/deps/requirements.py [hunk 19]
```diff
@@ -1,4 +1,11 @@
 import copy
+from enum import Enum
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/deps/requirements.py]:
```suggestion
```

File: lib/galaxy/tool_util/deps/requirements.py [hunk 20]
```diff
@@ -1,4 +1,11 @@
...
+from typing import (
+    Callable,
+    Dict,
+    Optional,
+    Union,
+)
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/deps/requirements.py:9]:
```suggestion
)

from typing_extensions import (
    get_args,
    Literal,
)

```

File: lib/galaxy/tool_util/deps/requirements.py [hunk 21]
```diff
@@ -252,10 +306,20 @@ def parse_requirements_from_xml(xml_root):
...
+        resource_elems = requirements_elem.findall("resource_requirement") if requirements_elem else []
+        resources = [resource_from_element(r) for r in resource_elems]
+        return requirements, containers, resources
 
     return requirements, containers
 
 
+def resource_from_element(resource_elem):
+    value_or_expression = xml_text(resource_elem)
+    resource_type = resource_elem.get("type", DEFAULT_CONTAINER_TYPE)
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/deps/requirements.py]:
```suggestion
    resource_type = resource_elem.get("type")
```

File: lib/galaxy/tool_util/deps/requirements.py [hunk 22]
```diff
@@ -204,20 +211,67 @@ def __str__(self):
...
+    ram_min = "ram_min"
+    ram_max = "ram_max"
+    tmpdir_min = "tmpdir_min"
+    tmpdir_max = "tmpdir_max"
+
+
+class ResourceRequirement:
+    def __init__(self, value_or_expression: Union[int, float, str], resource_type: ResourceType):
+        self.value_or_expression = value_or_expression
+        self.resource_type = resource_type
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/deps/requirements.py:238]:
```suggestion
        if not resource_type:
            raise ValueError("Missing resource requirement type")
        if resource_type not in VALID_RESOURCE_TYPES:
            raise ValueError(f"Invalid resource requirement type '{resource_type}'")
        self.resource_type = resource_type
```

File: lib/galaxy/tool_util/deps/requirements.py (hunk 19 above)
Comment (mvdbeek) [reply lib/galaxy/tool_util/deps/requirements.py]: What's the idea here ? Don't we need this for the ResourceType ?

File: lib/galaxy/tool_util/deps/requirements.py (hunk 20 above)
Comment (mvdbeek) [reply lib/galaxy/tool_util/deps/requirements.py:9]: Not using Literal or get_args ... I suppose there's a suggestion missing here ?

File: lib/galaxy/tool_util/deps/requirements.py (hunk 22 above)
Comment (mvdbeek) [reply lib/galaxy/tool_util/deps/requirements.py:238]: Again, `VALID_RESOURCE_TYPES` missing ?

File: test/functional/tools/resource_requirements.xml (hunk 18 above)
Comment (mvdbeek) [reply test/functional/tools/resource_requirements.xml:3]: For Xml tools yes, for CWL and YAML I don't think they should go under requirements.

Comment (mvdbeek) [reply test/functional/tools/resource_requirements.xml:3]: Hmm, okay, no, we should put them under requirement

Comment (mvdbeek) [reply test/functional/tools/resource_requirements.xml:3]: for cwl ... I guess for yaml tools the idea was to have additional requirement types at the root level, like we do for containers. not sure that was a good idea though

File: lib/galaxy/tool_util/deps/requirements.py [hunk 23]
```diff
@@ -204,20 +211,67 @@ def __str__(self):
         return f"ContainerDescription[identifier={self.identifier},type={self.type}]"
 
 
+class ResourceType(str, Enum):
+    cores_min = "cores_min"
+    cores_max = "cores_max"
+    ram_min = "ram_min"
+    ram_max = "ram_max"
+    tmpdir_min = "tmpdir_min"
+    tmpdir_max = "tmpdir_max"
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/deps/requirements.py]:
```suggestion
ResourceType = Literal[
    "cores_min",
    "cores_max",
    "ram_min",
    "ram_max",
    "tmpdir_min",
    "tmpdir_max",
]
VALID_RESOURCE_TYPES = get_args(ResourceType)
```

File: lib/galaxy/tool_util/deps/requirements.py (hunk 22 above)
Comment (nsoranzo) [reply lib/galaxy/tool_util/deps/requirements.py:238]: Mmmh, one suggestion got lost somehow, re-added now.

File: lib/galaxy/tool_util/deps/requirements.py [hunk 24]
```diff
@@ -211,18 +216,24 @@ def __str__(self):
...
 
 
-class ResourceType(str, Enum):
-    cores_min = "cores_min"
-    cores_max = "cores_max"
-    ram_min = "ram_min"
-    ram_max = "ram_max"
-    tmpdir_min = "tmpdir_min"
-    tmpdir_max = "tmpdir_max"
+ResourceType = Literal[
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/deps/requirements.py:220]: How do I use this with non-literal values ? Why is this better than a Enum ?

Comment (nsoranzo) [reply lib/galaxy/tool_util/deps/requirements.py:220]: It's currently only used for static checking, what do you have in mind?

Comment (mvdbeek) [reply lib/galaxy/tool_util/deps/requirements.py:220]: https://github.com/galaxyproject/galaxy/pull/5378/commits/6972f0ffad996257065bc578992cde25f38126b9

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 25]
```diff
@@ -680,6 +691,17 @@ resolver.
       </xs:extension>
     </xs:simpleContent>
   </xs:complexType>
+  <xs:complexType name="Resource">
+    <xs:simpleContent>
+      <xs:extension base="xs:string">
+        <xs:attribute name="type" type="ResourceType" use="required">
+          <xs:annotation>
+            <xs:documentation xml:lang="en">This value describes the resource available to the tool at runtime. Currently not implemented in Galaxy.</xs:documentation>
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
            <xs:documentation xml:lang="en">This value describes the type of resource required by the tool at runtime. Not yet implemented in Galaxy.</xs:documentation>
```

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 26]
```diff
@@ -598,11 +598,22 @@ practice.
     </xs:attribute>
   </xs:complexType>
 
+  <xs:complexType name="ResourceRequirement">
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]: Is this still used?

File: lib/galaxy/tool_util/xsd/galaxy.xsd [hunk 27]
```diff
@@ -598,11 +598,22 @@ practice.
...
+    <xs:sequence>
+      <xs:element name="cores_min" type="Resource" minOccurs="0" maxOccurs="1" />
+      <xs:element name="cores_max" type="Resource" minOccurs="0" maxOccurs="1" />
+      <xs:element name="ram_min" type="Resource" minOccurs="0" maxOccurs="1" />
+      <xs:element name="ram_max" type="Resource" minOccurs="0" maxOccurs="1" />
+      <xs:element name="tmpdir_min" type="Resource" minOccurs="0" maxOccurs="1" />
+      <xs:element name="tmpdir_max" type="Resource" minOccurs="0" maxOccurs="1" />
+    </xs:sequence>
+  </xs:complexType>
+
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/xsd/galaxy.xsd]:
```suggestion
```

File: lib/galaxy/tool_util/deps/requirements.py [hunk 28]
```diff
@@ -204,20 +217,97 @@ def __str__(self):
...
+
+    @property
+    def runtime_required(self):
+        if self._runtime_required is None:
+            try:
+                float(self.value_or_expression)
+                self._runtime_required = False
+            except ValueError:
+                self._runtime_required = True
+        return self._runtime_required
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/deps/requirements.py]: Can we simplify here?

```suggestion
        try:
            float(self.value_or_expression)
            self.runtime_required = False
        except ValueError:
            self.runtime_required = True
```

File: lib/galaxy/tool_util/parser/interface.py [hunk 29]
```diff
@@ -182,7 +182,7 @@ def parse_required_files(self) -> Optional["RequiredFiles"]:
 
     @abstractmethod
     def parse_requirements_and_containers(self):
-        """Return pair of ToolRequirement and ContainerDescription lists."""
+        """Return triple of ToolRequirement, ContainerDescription and ResourceRequirements lists."""
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/parser/interface.py]:
```suggestion
        """Return triple of ToolRequirement, ContainerDescription and ResourceRequirement lists."""
```

File: lib/galaxy/util/xml_macros.py [hunk 30]
```diff
@@ -5,10 +5,12 @@
...
-from galaxy.util import (
-    Element,
-    parse_xml,
-)
+from galaxy.util import parse_xml
+
+try:
+    from lxml.etree import _Element
+except ImportError:
+    pass
```
Comment (nsoranzo) [inline lib/galaxy/util/xml_macros.py]: What error is aabac58533fdb5aa2da7dd679a562aad370c9bb3 fixing exactly? The galaxy-util package doesn't require lxml currently, so mypy should fail??

Comment (mvdbeek) [reply lib/galaxy/util/xml_macros.py]: Hmm, mypy was complaining that you can't use a variable for type annotations ... but I reverted the commit and mypy doesn't fail, so ... 🤷

**Review by nsoranzo — APPROVED** [review]

File: lib/galaxy/util/xml_macros.py (hunk 30 above)
Comment (mvdbeek) [reply lib/galaxy/util/xml_macros.py]: Alright, seeing that again locally:
```
lib/galaxy/util/xml_macros.py:29: error: Variable "galaxy.util.Element" is not valid as a type
[valid-type]
        macros: Dict[str, List[Element]] = {}
                               ^
lib/galaxy/util/xml_macros.py:29: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#variables-vs-type-aliases
lib/galaxy/util/xml_macros.py:36: error: Element? has no attribute "get"  [attr-defined]
            tokens[m.get("name")] = m.text or ""
                   ^
lib/galaxy/util/xml_macros.py:36: error: Element? has no attribute "text"  [attr-defined]
            tokens[m.get("name")] = m.text or ""
                                    ^
lib/galaxy/util/xml_macros.py:42: error: Element? has no attribute "get"  [attr-defined]
            macro_dict[m.get("name")] = XmlMacroDef(m)
                       ^
Found 4 errors in 1 file (checked 1460 source files)
```

Comment (nsoranzo) [reply lib/galaxy/util/xml_macros.py]: Probably some different package versions w.r.t. lib/galaxy/dependencies/pinned-lint-requirements.txt ?

Comment (mvdbeek) [reply lib/galaxy/util/xml_macros.py]: absolutely possible, but it also makes sense, doesn't it ? What do you think about https://github.com/galaxyproject/galaxy/pull/14197 ?

**Pull request merged**